### PR TITLE
Initialize VarHandle.methodHandleTable and its entries with atomic CAS

### PR DESCRIPTION
**Pull request by jdmpapin**
This prevents double initialization when multiple threads are racing through `getMethodHandle()`. Once a non-null reference has been written to one of these locations, any subsequent write to the same location is a problem for the OpenJ9 JIT given its current handling of known objects.

Issue: eclipse-openj9/openj9#18882

### CODE REVIEW

File: src/java.base/share/classes/java/lang/invoke/VarHandle.java [hunk 1]
```diff
@@ -2198,15 +2206,32 @@ public Optional<VarHandleDesc> describeConstable() {
...
     final MethodHandle getMethodHandle(int mode) {
         MethodHandle[] mhTable = methodHandleTable;
         if (mhTable == null) {
-            mhTable = methodHandleTable = new MethodHandle[AccessMode.COUNT];
+            mhTable = new MethodHandle[AccessMode.COUNT];
+            if (!UNSAFE.compareAndSetReference(this, METHOD_HANDLE_TABLE_OFFSET, null, mhTable)) {
+                // We lost the race. Use the winning table instead.
+                mhTable = methodHandleTable;
+                assert mhTable != null;
+            }
```
Comment (keithc-ca) [inline src/java.base/share/classes/java/lang/invoke/VarHandle.java:2224]: Suggest this should use `compareAndExchangeReference()` which avoids an unnecessary access and makes it clear the result is non-null:
```java
            Object other = UNSAFE.compareAndExchangeReference(this, METHOD_HANDLE_TABLE_OFFSET, null, mhTable);
            if (other != null) {
                // We lost the race. Use the winning table instead.
                mhTable = (MethodHandle[])other;
            }
```
Likewise for element access below:
```java
            long offset = Unsafe.ARRAY_OBJECT_BASE_OFFSET + (Unsafe.ARRAY_OBJECT_INDEX_SCALE * mode);
            Object other = UNSAFE.compareAndExchangeReference(mhTable, offset, null, mh);
            if (other != null) {
                // We lost the race. Use the winning table instead.
                mh = (MethodHandle)other;
            }
```

File: src/java.base/share/classes/java/lang/invoke/VarHandle.java [hunk 2]
```diff
@@ -23,8 +23,16 @@
...
 
+/*
+ * ===========================================================================
+ * (c) Copyright IBM Corp. 2024, 2024 All Rights Reserved
+ * ===========================================================================
+ */
+
 package java.lang.invoke;
 
+import jdk.internal.misc.Unsafe;
```
Comment (keithc-ca) [inline src/java.base/share/classes/java/lang/invoke/VarHandle.java:34]: Please put this import with others for jdk.internal.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

**Comment by keithc-ca** [comment]
> a problem for the OpenJ9 JIT given its current handling of known objects

I'm not sure the problem is confined to OpenJ9: the assertion is generated by [InvokerBytecodeGenerator](https://github.com/ibmruntimes/openj9-openjdk-jdk/blob/openj9/src/java.base/share/classes/java/lang/invoke/InvokerBytecodeGenerator.java#L720).

**Comment by jdmpapin** [comment]
Yes, but a subsequent write only causes the assertion to fail because of the way we deal with known objects. Otherwise, we'd have gotten the LF from the MH at runtime and it would match

**Comment by keithc-ca** [comment]
> Yes, but a subsequent write only causes the assertion to fail because of the way we deal with known objects. Otherwise, we'd have gotten the LF from the MH at runtime and it would match

I'm not sure whether to infer from that response that future changes to OpenJ9 are anticipated which will eliminate the need for this patch.

**Comment by jdmpapin** [comment]
Yes, we anticipate future OpenJ9 changes will allow this to be reverted, in particular the implementation of https://github.com/eclipse-openj9/openj9/issues/16616. See https://github.com/eclipse-openj9/openj9/issues/18882#issuecomment-1958173709 and https://github.com/eclipse-openj9/openj9/issues/18882#issuecomment-1960277863. It will be some time though. I have some not-yet-contributed partial progress toward it, but none of that is on the code generation / GC interactions, which I expect to be the bulk of the work

**Comment by jdmpapin** [comment]
Updated to address review comments

**Comment by keithc-ca** [comment]
Jenkins test sanity.openjdk zlinux jdknext

**Comment by keithc-ca** [comment]
Jenkins copyright check

**Comment by keithc-ca** [comment]
The failures seem to only be with respect to `java.lang.Class.toGenericString()` because `sealed` is missing; see https://github.com/eclipse-openj9/openj9/issues/19033.

**Review by keithc-ca — APPROVED** [review]

**Pull request merged**